### ISSUE TITLE
Depthwise Convolution Gradient Shape Mismatch in Reverse Mode Autodiff

### ISSUE THREAD

**Issue by Guru2308** [gomlx/gomlx#132]
When using **depthwise convolution** in `gomlx`, the gradient computed for the input during backpropagation does not match the expected shape. Specifically, the accumulated VJP (gradient) for the input `x` is incorrectly computed with shape `[batch, H, W, 1]` instead of `[batch, H, W, C]`, leading to a mismatch in subsequent operations.

**Steps to Reproduce**

1. Define a **Depthwise Convolution** layer using:
    
    ```go
    convOpts := graph.Convolve(x, kernel).
        ChannelsAxis(channelsAxisConfig).
        StridePerDim(strides...).
        FilterGroupCount(inputChannels). // Depthwise mode
        PadSame()
    ```
    
2. During training, backpropagation fails with:
    
    ```
    invalid shape (or DType) for calculated AccumulatedVJP for input #0:
    input shape=(Float32)[128 8 8 128],
    calculated AccumulatedVJP shape=(Float32)[128 8 8 1]
    (wanted (Float32)[128 8 8 128])
    ```
    
3. The issue originates in `convGeneralDilatedVJP` where `convVJPWrtX` incorrectly computes `vjpX`.

**Expected Behaviour**

- The computed gradient should have the same shape as the input, i.e., `[batch, H, W, C]`, rather than `[batch, H, W, 1]`.
- `convVJPWrtX` should correctly reverse the grouped convolution by broadcasting the gradient across all input channels.

**Environment**

- **gomlx version**: `v0.6.0`
- **Go version**: `1.23.6`
- **Backend**: XLA (CPU)

**Additional Information**

- The error message in `rev_autodiff.go` suggests that this might be an **internal bug**, as it explicitly states:
    
    > "this probably indicates a bug in the code, please report the issue."
    
- This issue occurs when using `FilterGroupCount(inputChannels)`, which activates **depthwise mode**.
- The **forward pass works correctly**, but the **gradient calculation fails**.

Would appreciate any insights on whether this fix is appropriate or if there's a better way to handle grouped convolutions in reverse-mode autodiff.

**Comment by janpfeifer** [comment]
hi @Guru2308 , thanks for the detailed report! 

While implementing the convolution back-propagation, I didn't cover all the variations, and this may have been one of the missing ones. Let me see if I can find a quick fix here.

**Comment by janpfeifer** [comment]
It's been a while since I wrote this part, but now I recall never having exposed the  `featureGroupCount` in the `Convolve(x, kernel)` API.  There is no `ConvolutionBuilder.FeatureGroupCount()` method (yet). Neither is `batchGroupCount`.

Important: one can do depth-wise convolution without `FeatureGroupCount`. Just set the output channels dimension of your kernel accordingly. They should be shaped `[<spatial_dimensions...>, input_channels, output_channels]` (you can change the order with `ChannelsAxis()`). Just set the `output_channels` to the desired "depth" of the convolution, if you will.

Question: do you need specifically `FeatureGroupCount` ? It will take me a while to implement (this month my time availability is more limited), but I'll get to it.

Also, if yes, an option in the short term that may work for you is to simply set `output_channels = input_channels * channels_multiplier`. This will use many more parameters (so it's an overkill), but will be "more powerful" than the `featureGroupCount` -- but also "more overfitting".

In the meantime I added an better error message about not supporting `featureGroupCount` and `batchGroupCount` in the back-propagation (a.k.a. "reverse auto-grad"). It's submitted in HEAD (`main` branch).


ps.: GoMLX is already in v0.6.2, just FYI.

**Comment by janpfeifer** [comment]
For the record, documentation on `featureGroupCount` (from Tensorflow): https://www.tensorflow.org/api_docs/python/tf/nn/depthwise_conv2d

**Comment by janpfeifer** [comment]
Marking this as _enchancement_ (a.k.a. "feature request") for now: adding `FeatureGroupCount` support.

I'm planning to get to this next month.

**Comment by Guru2308** [comment]
> Marking this as _enchancement_ (a.k.a. "feature request") for now: adding `FeatureGroupCount` support.
> 
> I'm planning to get to this next month.

Actually I've been working on this. When I started working on the Separable layer, i started with this as well

**Comment by janpfeifer** [comment]
> > Marking this as _enchancement_ (a.k.a. "feature request") for now: adding `FeatureGroupCount` support.
> > I'm planning to get to this next month.
> 
> Actually I've been working on this. When I started working on the Separable layer, i started with this as well. You can check my fork and lemme know if I'm doing it properly

Oh, cool, looking forward to it @Guru2308 ! Thanks for doing it.

**Comment by Guru2308** [comment]
> > > Marking this as _enchancement_ (a.k.a. "feature request") for now: adding `FeatureGroupCount` support.
> > > I'm planning to get to this next month.
> > 
> > 
> > Actually I've been working on this. When I started working on the Separable layer, i started with this as well. You can check my fork and lemme know if I'm doing it properly
> 
> Oh, cool, looking forward to it [@Guru2308](https://github.com/Guru2308) ! Thanks for doing it. Send me the PR when you have it 😃

I've just added the FeatureGroupCount function to the `ops_convolve` and didnt make any changes to the forward/back propagation part. Is it okay just to raise PR with it? Once you fix the propagation part, I'll raise the PR for separable convolution layer.

**Comment by janpfeifer** [comment]
Yes, that works! If you add in `ops_convolve.go` it will already work for the forward pass. A couple of nice to have:

1. add a trivial test that the forward pass works -- like the others already there.
2. add a comment in the method that the backward propagation is not working yet. We can remove that comment later.

cheers

**Comment by Guru2308** [comment]
I've raised the PR, lemme know if that's okay.

**Comment by Guru2308** [comment]
I've raised the PR with the mentioned changes. Please check and lemme know :)

**Comment by janpfeifer** [comment]
I saw it, thank you @Guru2308 , reviewing it now

**Comment by Guru2308** [comment]
Thank you so much for letting me know. I think now I can work on the separable convolution layer without any issues.
